### DISCUSSION TITLE
Why does vfile-matter 4.0.0 use the  yaml library to parse Yaml front matter?

### DISCUSSION THREAD

**Discussion by bfdes** [remarkjs/remark#1006] [Q&A] [answered]
I'd like to know why [nodeca/js-yaml](https://github.com/nodeca/js-yaml) was replaced with [eemeli/yaml](https://github.com/eemeli/yaml) in https://github.com/vfile/vfile-matter/commit/1bd1da24cb35982bd6915b9fcb0317895250fa1c. I can't find any discussion in [vfile-matter](https://github.com/vfile/vfile-matter)'s issue history.

It turns out that https://github.com/eemeli/yaml follows the Yaml 1.2 spec in such a way that dates are unsupported. See https://github.com/eemeli/yaml/issues/132. In practice, this means Jekyll front matter like 

```
---
title: Property testing
tags: [Testing]
created: 2018-10-15
---
```

cannot be parsed out of the box.

I can work around the problem by asking the plugin to follow the Yaml 1.1 spec.

**Comment by ChristianMurphy** [comment]
It was changed in https://github.com/unifiedjs/unified-engine/issues/65 and has rolled out elsewhere since, there is further discussion in the linked issues.

/cc @remcohaszing

**Comment by wooorm** [comment]
Your problem seems to be that you want to use yaml 1.1.
You can pass that as an option: https://github.com/vfile/vfile-matter#optionsyaml

**Comment by remcohaszing** [comment, marked as answer]
The choice to switch unified projects was made when I simply DM’ed @wooorm about replacing [`js-yaml`](https://github.com/nodeca/js-yaml) with [`yaml`](https://github.com/eemeli/yaml), because `js-yaml` isn’t fully spec compliant. He said he had also run into this issue and immediately agreed, so no discussion was needed.

However, it’s good you asked why the change was made. I’ll elaborate why the change was made and why you should too in your projects. :smile: 

The YAML specification can be found [here](https://yaml.org/spec/).
A nice overview of changes in the YAML specification can be found [here](https://yaml.org/spec/1.2.2/ext/changes/#changes-in-version-12-revision-120-2009-07-21).

As can be seen in the YAML specification changelog, some significant changes were made in YAML 1.2 in 2009, notably:

> - Only `true` and `false` strings are parsed as booleans (including `True` and `TRUE`); `y`, `yes`, `on`, and their negative counterparts are parsed as strings.
> - Underlines `_` cannot be used within numerical values.
> - Octal values need a 0o prefix; e.g. `010` is now parsed with the value 10 rather than 8.
> - The binary and sexagesimal integer formats have been dropped.
> - The `!!pairs`, `!!omap`, `!!set`, `!!timestamp` and `!!binary` types have been dropped.
> - The merge `<<` and value `=` special mapping keys have been removed.

Let’s see how different projects handle this. Create a new directory and inside, run:

```sh
npm install yaml@2 js-yaml@4
```

Let’s create a script named `script.mjs`:

```js
import yaml from 'yaml'
import jsYaml from 'js-yaml'

const data = `
yep: yes
nope: no
date: 2022-06-15 12:00:00+02:00
number: 010
anchor: &anchor
  anchorObjectKey: anchorObjectValue
objectWithKeys:
  <<: *anchor
`

console.log('yaml parse:')
console.dir(yaml.parse(data))
console.log()

console.log('yaml parse 1.1:')
console.dir(yaml.parse(data, { version: '1.1' }))
console.log()

console.log('js-yaml load:')
console.dir(jsYaml.load(data))
console.log()
```

This outputs the following, except I added comments manually:

```js
yaml parse:
{
  // Parsed as string, as per YAML specification 1.2
  yep: 'yes',
  // Parsed as string, as per YAML specification 1.2
  nope: 'no',
  // Parsed as string, as per YAML specification 1.2
  date: '2022-06-15 12:00:00+02:00',
  // Parsed as 10, as per YAML specification 1.2
  number: 10,
  anchor: { anchorObjectKey: 'anchorObjectValue' },
  // << is a regular key, as per YAML specification 1.2
  objectWithKeys: { '<<': { anchorObjectKey: 'anchorObjectValue' } }
}

yaml parse 1.1:
{
  // Parsed as boolean, as per YAML specification 1.1
  yep: true,
  // Parsed as boolean, as per YAML specification 1.1
  nope: false,
  // Parsed as date, as per YAML specification 1.1.
  // Note that timezone information is lost when converting to JavaScript Date objects
  date: 2022-06-15T10:00:00.000Z,
  // Parsed as 8, as per YAML specification 1.1
  number: 8,
  anchor: { anchorObjectKey: 'anchorObjectValue' },
  // << is a merge key, as per YAML specification 1.1
  objectWithKeys: { anchorObjectKey: 'anchorObjectValue' }
}

js-yaml load:
{
  // Parsed as string, as per YAML specification 1.2
  yep: 'yes',
  // Parsed as string, as per YAML specification 1.2
  nope: 'no',
  // Parsed as date, as per YAML specification 1.1.
  // Note that timezone information is lost when converting to JavaScript Date objects
  date: 2022-06-15T10:00:00.000Z,
  // Parsed as 10, as per YAML specification 1.2
  number: 10,
  anchor: { anchorObjectKey: 'anchorObjectValue' },
  // << is a merge key, as per YAML specification 1.1
  objectWithKeys: { anchorObjectKey: 'anchorObjectValue' }
}
```

This shows that by default, the `yaml` library is compliant with YAML 1.2, and can be compliant with YAML 1.1 with a single option, or by specifying the yaml version using:

```yaml
%YAML 1.1
---
actual: data
```

The `js-yaml` library however, mixes the two, which leads to confusion and incompatibility.

Apart from spec compliancy, `yaml` also contains some cool features that `js-yaml` doesn’t and most people probably don’t care about:

- `yaml` ships type definitions.
- `yaml` contains a powerful and easy-to-use AST API. This allows you to modify YAML documents without the loss of formatting, comments, or anchors.
- `yaml` is written with JavaScript in mind from the start, whereas `js-yaml` is a port of `pyyaml`. This can be seen from the fact `yaml` uses `parse` and `stringify`, named after `JSON.parse` and `JSON.stringify`, whereas `js-yaml` uses `load` and `dump`, named after the [Python `json` module](https://docs.python.org/3/library/json.html).
- [`yaml-language-server`](https://github.com/redhat-developer/yaml-language-server) moved from `js-yaml` to `yaml`, which significantly increased the editor experience for YAML files.

**Reply by remcohaszing** [reply]
I would like to add I don’t mean to bash on `js-yaml`. It’s a great library that has served the JavaScript community for many years, but nowadays I believe `yaml` is the better option.

**Reply by bfdes** [reply]
It looks like js-yaml's README says it _started_ as a port of py-yaml. The library was rewritten since.

**Comment by bfdes** [comment]
Thanks for the comprehensive answers. As a library consumer, it isn't always obvious why dependencies change.

I'll try to switch to 1.2 (rather than forcing the plugin to conform to the 1.1 spec).
